File: src/levels/L12-chain/strategy.rb
# region: Clerk#approve
class Clerk
  def approve(request) = @refunds.pay(request)
end
# endregion

# region: Manager#approve
class Manager
  def approve(request) = @refunds.pay(request)
end
# endregion

# region: Owner#approve
class Owner
  def approve(request) = @refunds.pay(request)
end
# endregion

class Cashier
  APPROVERS = { bajo: Clerk, medio: Manager, alto: Owner }

  # Elegir aprobador por categoría funciona… mientras sepas de antemano todas las categorías.
  # region: Cashier#refund
  def refund(request) = APPROVERS.fetch(request.category).new.approve(request) # KeyError con montos enormes
  # endregion
end

File: src/levels/L12-chain/ts/strategy.ts
// region: Clerk#approve
class Clerk {
  private refunds = new Refunds()
  approve(request: RefundRequest) { this.refunds.pay(request) }
}
// endregion

// region: Manager#approve
class Manager {
  private refunds = new Refunds()
  approve(request: RefundRequest) { this.refunds.pay(request) }
}
// endregion

// region: Owner#approve
class Owner {
  private refunds = new Refunds()
  approve(request: RefundRequest) { this.refunds.pay(request) }
}
// endregion

class Cashier {
  static APPROVERS: Record<string, new () => { approve(request: RefundRequest): void }> = {
    bajo: Clerk, medio: Manager, alto: Owner,
  }

  // Elegir aprobador por categoría funciona… mientras sepas de antemano todas las categorías.
  // region: Cashier#refund
  refund(request: RefundRequest) { new Cashier.APPROVERS[request.category]().approve(request) } // TypeError con montos enormes
  // endregion
}
